EMTF DQM Module

swang373 opened the pull request:
#### EMTF DQM Module
- L1TStage2EMTF.h
- L1TStage2EMTF.cc
- L1TStage2EMTF_cfi.py

This module was written following the naming convention set forth by Jing Yu and Esmaeel and currently monitors basic hardware values stored in the EMTFOutputCollection.
#### EMTF Unpacker

emtfStage2Digis_cfi.py
EMTFOutput.h

The file EMTFOutput.h was modified with permission from Andrew to fix some method declarations. The python configuration file was provided by Andrew, with some naming changes to follow convention.
#### Testing

This module was compiled and tested within CMSSW_8_0_0_pre6. The input files used were the following from MWGR2:
- 'root://eoscms//store/express/Commissioning2016/ExpressCosmics/FEVT/Express-v1/000/264/593/00000/E8F68B40-0ED2-E511-90DB-02163E0145B3.root'
- 'root://eoscms//store/express/Commissioning2016/ExpressCosmics/FEVT/Express-v1/000/264/593/00000/EA333C9D-2BD2-E511-AF4D-02163E0138BE.root'
- 'root://eoscms//store/express/Commissioning2016/ExpressCosmics/FEVT/Express-v1/000/264/593/00000/EC0A64B5-21D2-E511-B525-02163E0136D0.root'
- 'root://eoscms//store/express/Commissioning2016/ExpressCosmics/FEVT/Express-v1/000/264/593/00000/EC871D40-1AD2-E511-96AF-02163E0133C2.root'

swang373 committed:
First commit, compiles and runs

cmsbuild commented:
A new Pull Request was created by  @swang373 (Sean-Jiun Wang) for CMSSW_8_0_X.

It involves the following packages:

DQM/L1TMonitor
DataFormats/L1TMuon
EventFilter/L1TRawToDigi


The following packages do not have a category, yet:

DataFormats/L1TMuon

@cmsbuild, @vanbesien, @mulhearn, @deguio, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

mulhearn commented:
I'm OK with this, but note I haven't turned on EMTF unpacking in L1T standard sequences yet.  Indeed, I'd like to leave this out of MWGR until we have exercised the L1T+HLT without any glitches, and have solid integration test with RAW data file that includes all of Stage-2 L1T.  Of course, we can have special version of sequence for online DQM... just talking about what goes in release.

mulhearn commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 27c069c8d21efe63707169d1c110f14c4a902987
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13513/11607/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13513/11607/summary.html

deguio commented:
please submit to 81x as well.

deguio commented:
ciao swang373,
I am reviewing the code right now. how are we supposed to run this code? do you plan to include the modules from the stage2 config at:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DQM/Integration/python/clients/l1tstage2_dqm_sourceclient-live_cfg.py

or do provide an additional cfg? including esmaeel in this thread.


please let us know asap, cause at the moment we cannot proceed (unless I am missing something).


@EsmaeelEskandari @jingyucms

swang373 commented:
Ciao Federico,

You are correct, I had not modified the necessary configuration files in DQM/L1TMonitor. Sorry about that.

In https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DQM/L1TMonitor/python/L1TStage2_cff.py:
The following imports are necessary
```python
from DQM.L1TMonitor.L1TStage2EMTF_cfi import *
from EventFilter.L1TRawToDigi.emtfStage2Digis_cfi import *
```

and l1tStage2Emtf needs to be added to the l1tStage2online sequence.

Then the emtfStage2Digis need to be added in the main configuration file on line 55: https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DQM/Integration/python/clients/l1tstage2_dqm_sourceclient-live_cfg.py#L55

This was enough for me to test the code locally with an l1tstage2_cfg file Esmaeel provided me.

deguio commented:
ok. could you add a commit to this very same PR with the needed changes?
once done please don't forget to submit the same PR to 81x as well.
thanks,
F.

swang373 committed:
Necessary configuration files modified.

The configuration files were modified heavily to
test the ability to run locally on lxplus, then
the settings were reverted, keeping only those
changes which were relevant to enable running the
EMTF DQM module.

cmsbuild commented:
Pull request #13513 was updated. @cmsbuild, @vanbesien, @mulhearn, @deguio, @davidlange6 can you please check and sign again.

swang373 commented:
Ciao Federico,

In my CMSSW_8_0_0_pre6 release area, the versions of the necessary configuration files were older than the ones you pointed me too. I've tried merging and some other ideas, but they didn't work and I'm convinced I don't know the workflow for developing CMSSW with git as well as I should.

I downloaded the raw code for the configuration file and overwrote the older version, then made the necessary changes to test running on lxplus (uncommenting relevant lines, etc.) Before committing, I reverted the changes I made to allow for running on lxplus and left the necessary modifications. I guess that still left some complications to be manually merged, so my apologies.

swang373 commented:
Also, pardon the naive question, but is there a particular CMSSW_8_1_X release that I should work in, or will any suffice? I'll make a PR for 81x as soon as I can.

deguio commented:
this now needs to be rebased.  :(

deguio commented:
ciao @swang373 @EsmaeelEskandari 
please let me know what is the ETA for this PR to be rebased and ready for testing. it would be nice to have it running over the weekend on our DQM playback system in P5.
cheers and thanks,
F.

swang373 commented:
Ciao @deguio ,

I'm aiming for ETA end of today at the latest. I've looked up how to rebase and I'll give that a try.

swang373 committed:
Add some whitespace to follow style

cmsbuild commented:
Pull request #13513 was updated. @cmsbuild, @vanbesien, @mulhearn, @deguio, @davidlange6 can you please check and sign again.

swang373 commented:
Ciao @deguio,

I've rebased the code and tested it within a CMSSW_8_0_0 release. With the necessary modifications of `l1tstage2_dqm_sourceclient-live_cfg.py` to run on lxplus, an output .root file is produced containing filled plots from the EMTF module as well as the CaloLayer2 and mGMT modules

deguio commented:
please test

deguio commented:
going to test this on the playback

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9acc5fda885b828cec5617e4d056c608043ce2cc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13513/11711/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7e4152f5e1dad4cbfbf20713b07c3f87825e4d5a
https://github.com/cms-sw/cmssw/commit/ef8ce2c331fa84061181315eaad9ca82d6cde23d
https://github.com/cms-sw/cmssw/commit/bd32fd670a8c81f02ae428e7b80b1e8639f254bf
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13513/11711/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13513/11711/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13513/11711/summary.html

deguio commented:
@swang373
please let me know which is the correspondent PR for 81x.
if you want to test, just choose a recent IB.
thanks,
F.

swang373 commented:
@deguio 
The correspondent PR for 81X is cms-sw/cmssw#13630.

deguio commented:
please test

deguio commented:
+1

mulhearn commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

deguio commented:
@swang373 
why this PR was closed? please re-open asap if you want it to be merged.
thanks,
F.

swang373 commented:
@deguio 

Sorry about that!

deguio commented:
no problem.
F.

davidlange6 commented on EventFilter/L1TRawToDigi/python/emtfStage2Digis_cfi.py:
```diff
@@ -0,0 +1,12 @@
+import FWCore.ParameterSet.Config as cms
```
what are the plans for also adding this unpacker to the tier-0? @mulhearn

davidlange6 commented on DataFormats/L1TMuon/interface/EMTFOutput.h:
```diff
@@ -50,27 +50,27 @@ namespace l1t {
-    emtf::MTF7Trailer GetMTF7Trailer()      { return MTF7Trailer;   };
-    emtf::AMC13Trailer GetAMC13Trailer()    { return AMC13Trailer;  };
-    const uint64_t Dataword() const     { return dataword; };
+    const bool HasAMC13Header()                  const { return hasAMC13Header;  };
```
why should the output be const for all of these functions? @swang373 @mulhearn

swang373 replied:
To be honest, I'm not sure. This appears to be following a similar convention to https://github.com/swang373/cmssw/blob/EMTF_DQM_dev/DataFormats/L1TMuon/interface/RegionalMuonCand.h. Perhaps @abrinke1 would be able to comment on this, as I implemented his fix in order for the EMTF DQM module to compile.

abrinke1 replied:
Not sure why they need to be const, but without it there were compile errors, and with it, it works.

davidlange6 replied:
i understand that the methods should be constant - but not their return types - what were the errors?

> On Mar 17, 2016, at 5:37 PM, abrinke1 <notifications@github.com> wrote:
> 
> In DataFormats/L1TMuon/interface/EMTFOutput.h:
> 
> > -    int  NumME()                            { return numME;           };
> > -    bool HasAMC13Trailer()                  { return hasAMC13Trailer; };
> > -    bool HasMTF7Trailer()                   { return hasMTF7Trailer;  };
> > -    bool HasEventTrailer()                  { return hasEventTrailer; };
> > -    emtf::AMC13Header GetAMC13Header()      { return AMC13Header;   };
> > -    emtf::MTF7Header GetMTF7Header()        { return MTF7Header;    };
> > -    emtf::EventHeader GetEventHeader()      { return EventHeader;   };
> > -    emtf::Counters GetCounters()            { return Counters;      };
> > -    emtf::MECollection GetMECollection()    { return MECollection;  };
> > -    emtf::RPCCollection GetRPCCollection()  { return RPCCollection; };
> > -    emtf::SPCollection GetSPCollection()    { return SPCollection;  };
> > -    emtf::EventTrailer GetEventTrailer()    { return EventTrailer;  };
> > -    emtf::MTF7Trailer GetMTF7Trailer()      { return MTF7Trailer;   };
> > -    emtf::AMC13Trailer GetAMC13Trailer()    { return AMC13Trailer;  };
> > -    const uint64_t Dataword() const     { return dataword; };
> > +    const bool HasAMC13Header()                  const { return hasAMC13Header;  };
> 
> Not sure why they need to be const, but without it there were compile errors, and with it, it works.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

swang373 replied:
Hi @davidlange6 , @abrinke1 ,

With further updates to the DataFormats for EMTF, we are looking into this comment to make sure we code it right. That the methods should be const is agreed upon. My understanding is that the edm::Handle returned by the call to getByToken() is const, so only const methods may be used to access its members.

That the return types do not have to be const, yes, I see that now. Pardon my confusion. If however, we are returning references or pointers to members, then the return type must be const. Is my understanding correct?
